File: app/models/rating.py
from datetime import datetime, timezone
from typing import Annotated, List, Optional
from beanie import Document, Indexed
from pydantic import Field


class Rating(Document):
    """MongoDB Document model for partner ratings and reviews submitted by customers."""

    order_id: Annotated[str, Indexed(unique=True)]  # One rating per order
    customer_id: Annotated[str, Indexed()]
    customer_name: Optional[str] = None
    partner_id: Annotated[str, Indexed()]  # User ID of the delivery partner
    rating: float = Field(..., ge=1.0, le=5.0, description="Star rating between 1.0 and 5.0")
    review: Optional[str] = Field(None, description="Optional customer review comment")
    tags: List[str] = Field(
        default_factory=list,
        description="Rating feedback tags, e.g. on_time, polite, careful_handling, fast_delivery",
    )
    is_hidden: bool = Field(False, description="Admin moderation flag to hide inappropriate reviews")
    admin_notes: Optional[str] = Field(None, description="Internal notes from admin regarding moderation")
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )

    class Settings:
        name = "ratings"
        indexes = [
            "order_id",
            "customer_id",
            "partner_id",
            [("partner_id", 1), ("is_hidden", 1), ("created_at", -1)],
            [("customer_id", 1), ("created_at", -1)],
            [("rating", -1)],
            [("created_at", -1)],
        ]

    def touch(self):
        """Update updated_at timestamp."""
        self.updated_at = datetime.now(timezone.utc)

File: src/models/Rating.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IRating extends Document {
  order_id: string;
  customer_id: string;
  customer_name?: string;
  partner_id: string;
  rating: number;
  review?: string;
  tags: string[];
  is_hidden: boolean;
  admin_notes?: string;
  created_at: Date;
  updated_at: Date;
  touch(): void;
}

const RatingSchema = new Schema<IRating>(
  {
    order_id: { type: String, required: true, unique: true, index: true },
    customer_id: { type: String, required: true, index: true },
    customer_name: { type: String },
    partner_id: { type: String, required: true, index: true },
    rating: { type: Number, required: true, min: 1, max: 5 },
    review: { type: String },
    tags: [{ type: String }],
    is_hidden: { type: Boolean, default: false },
    admin_notes: { type: String },
  },
  {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' },
  }
);

RatingSchema.index({ partner_id: 1, is_hidden: 1, created_at: -1 });
RatingSchema.index({ customer_id: 1, created_at: -1 });
RatingSchema.index({ rating: -1 });
RatingSchema.index({ created_at: -1 });

RatingSchema.methods.touch = function () {
  this.updated_at = new Date();
};

export const Rating = mongoose.model<IRating>('Rating', RatingSchema, 'ratings');
